moviebot/agent/agent.py

```diff
@@ -142,112 +135,117 @@ def _get_recommender(self, recommender_type: str) -> RecommenderModel:
-        agent_response, options = self.nlg.generate_output(
-            agent_dacts, user_fname=user_fname
-        )
-        self.dialogue_manager.get_context().add_utterance(
-            AgentUtterance(agent_response)
-        )
         if not self.isBot:
             logger.debug(
-                str(self.dialogue_manager.dialogue_state_tracker.dialogue_state)
+                str(
+                    self.dialogue_manager.dialogue_state_tracker.dialogue_state  # noqa
+                )
+            )
+            utterance = AnnotatedUtterance(
+                intent=agent_intent,
+                text=agent_response,
+                participant=DialogueParticipant.AGENT,
+                annotations=[],
+                metadata={"options": options},
             )
-            logger.debug(str(self.dialogue_manager.get_context()))
-            return agent_response, options
         else:
             record_data = self.dialogue_manager.get_state().to_dict()
-            context = self.dialogue_manager.get_context().movies_recommended
-            record_data.update(
-                {"Agent_Output": agent_response, "Context": context}
+            utterance = AnnotatedUtterance(
+                intent=agent_intent,
+                text=agent_response,
+                participant=DialogueParticipant.AGENT,
+                annotations=[],
+                metadata={"options": options, "record_data": record_data},
             )
-            return agent_response, options, record_data
 
-    def continue_dialogue(
-        self,
-        user_utterance: UserUtterance,
-        user_options: DialogueOptions,
-        user_fname: str = None,
-    ) -> Union[
-        Tuple[str, DialogueOptions],
-        Tuple[str, DialogueOptions, Dict[str, Any]],
-    ]:
-        """Performs the next dialogue according to user response and current
-        state of dialogue.
+        return utterance
+
+    def welcome(self, user_fname: str = None) -> None:
+        """Sends a welcome message to the user.
 
         Args:
-            user_utterance: The input received from the user.
-            user_options: Dialogue options that were provided to the user in
-              previous turn.
             user_fname: User's first name. Defaults to None.
+        """
+        agent_dacts = self.dialogue_manager.start_dialogue(self.new_user)
+        self.dialogue_manager.dialogue_state_tracker.update_state_agent(
+            agent_dacts
+        )
+        agent_response, options = self.nlg.generate_output(
+            agent_dacts, user_fname=user_fname
+        )
 
-        Returns:
-            Agent response, dialogue options for quick reply, and data to be
-              recorded if self.isBot is true.
+        utterance = self._generate_utterance(
+            agent_response, AgentIntents.WELCOME.value, options
+        )
+
+        self._dialogue_connector.register_agent_utterance(utterance)
+
+    def goodbye(self) -> None:
```

Comment:
I don't think this method is in use. Dialogue policy generates `bye` intent, and dialogue connector closes the conversation.

Reply:
Good catch. However, the method needs to be implemented as it is an abstract method in the parent class.